### PR TITLE
test: reproducible tests

### LINKED ISSUE #128: add macOS and windows testing to test.yml
resolved by #204

### LINKED ISSUE #202: test: make test cases (more) deterministic
See https://pytorch.org/docs/stable/notes/randomness.html
We'll also need something for the BMS module

### PR DESCRIPTION
**Pull request by hollandjg**
Make test cases involving neural networks more reproducible.

resolves #202 
resolves #128 
invalidates #204

Type of change:
- Testing update

### CODE REVIEW

**Review by listar2000 — COMMENTED** [review]
LGTM! The only question is should we allow inputting a custom `seed` instead of always using 180? A magic number should be OK if we only care about getting the same test.

**Comment by hollandjg** [comment]
> LGTM! The only question is should we allow inputting a custom `seed` instead of always using 180? A magic number should be OK if we only care about getting the same test.

Thanks @listar2000! 

For the test cases, all we care about is that they're reproducible when we run them in the GitHub actions runner. People shouldn't be using `seed=180` by default in their production code. I'll add a docstring in the test case runners. 

Some tests might be different – later we'll do more statistical tests about the convergence of the models in "productive" use, where we'll potentially want a totally different kind of seeding, so I don't see a good way to make this parametrizable right now.

I chose 180 for "180 George St.", the CCV office address. The alternative would have been 42.

**Review by listar2000 — APPROVED** [review]
Thanks @hollandjg for the explanation. That makes total sense and it LGTM.

**Comment by benwandrew** [comment]
> > LGTM! The only question is should we allow inputting a custom `seed` instead of always using 180? A magic number should be OK if we only care about getting the same test.
> 
> Thanks @listar2000!
> 
> For the test cases, all we care about is that they're reproducible when we run them in the GitHub actions runner. People shouldn't be using `seed=180` by default in their production code. I'll add a docstring in the test case runners.
> 
> Some tests might be different – later we'll do more statistical tests about the convergence of the models in "productive" use, where we'll potentially want a totally different kind of seeding, so I don't see a good way to make this parametrizable right now.
> 
> I chose 180 for "180 George St.", the CCV office address. The alternative would have been 42.

42 as in [The Answer to the Ultimate Question of Life, the Universe, and Everything](https://en.wikipedia.org/wiki/Phrases_from_The_Hitchhiker%27s_Guide_to_the_Galaxy#The_Answer_to_the_Ultimate_Question_of_Life,_the_Universe,_and_Everything_is_42)?

**Review by benwandrew — APPROVED** [review]
👍

**Comment by hollandjg** [comment]
> > > LGTM! The only question is should we allow inputting a custom `seed` instead of always using 180? A magic number should be OK if we only care about getting the same test.
> > 
> > 
> > Thanks @listar2000!
> > For the test cases, all we care about is that they're reproducible when we run them in the GitHub actions runner. People shouldn't be using `seed=180` by default in their production code. I'll add a docstring in the test case runners.
> > Some tests might be different – later we'll do more statistical tests about the convergence of the models in "productive" use, where we'll potentially want a totally different kind of seeding, so I don't see a good way to make this parametrizable right now.
> > I chose 180 for "180 George St.", the CCV office address. The alternative would have been 42.
> 
> 42 as in [The Answer to the Ultimate Question of Life, the Universe, and Everything](https://en.wikipedia.org/wiki/Phrases_from_The_Hitchhiker%27s_Guide_to_the_Galaxy#The_Answer_to_the_Ultimate_Question_of_Life,_the_Universe,_and_Everything_is_42)?

Exactly, "What do you get if you multiply six by nine?"
